meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java

```diff
@@ -56,56 +69,82 @@ public StoreType getStoreType() {
+    _connectionInitTimeoutInMillis = connectionInitTimeoutInMillis;
+    _sessionTimeoutInMillis = sessionTimeoutInMillis;
     _enableAuth = enableAuth;
     _storeType = storeType;
   }
 
-  public static class Builder {
-    private String _connectionAddress;
-
-    private long _connectionTimeout;
-    private boolean _enableAuth;
-    //private RetryProtocol _retryProtocol;
-    private StoreType _storeType;
+  public static class MetaClientConfigBuilder<B extends MetaClientConfigBuilder<B>> {
+    protected String _connectionAddress;
 
+    protected long _connectionInitTimeoutInMillis;
+    protected long _sessionTimeoutInMillis;
+    // protected long _operationRetryTimeout;
+    // protected RetryProtocol _retryProtocol;
+    protected boolean _enableAuth;
+    protected StoreType _storeType;
 
 
     public MetaClientConfig build() {
       validate();
-      return new MetaClientConfig(_connectionAddress, _connectionTimeout, _enableAuth, _storeType);
+      return new MetaClientConfig(_connectionAddress, _connectionInitTimeoutInMillis,
+          _sessionTimeoutInMillis,
+          _enableAuth, _storeType);
     }
 
-    public Builder() {
+    public MetaClientConfigBuilder() {
       // set default values
       setAuthEnabled(false);
-      setConnectionTimeout(-1);
+      setConnectionInitTimeoutInMillis(MetaClientConstants.DEFAULT_CONNECTION_INIT_TIMEOUT_MS);
+      setSessionTimeoutInMillis(MetaClientConstants.DEFAULT_SESSION_TIMEOUT_MS);
     }
 
-    public Builder setConnectionAddress(String connectionAddress) {
+    public B setConnectionAddress(String connectionAddress) {
       _connectionAddress = connectionAddress;
-      return this;
+      return self();
     }
 
-    public Builder setAuthEnabled(Boolean enableAuth) {
+    public B setAuthEnabled(Boolean enableAuth) {
       _enableAuth = enableAuth;
-      return this;
+      return self();
+    }
+
+    /**
+     * Set timeout in mm for connection initialization timeout
+     * @param timeout
+     * @return
+     */
+    public B setConnectionInitTimeoutInMillis(long timeout) {
+      _connectionInitTimeoutInMillis = timeout;
+      return self();
     }
 
-    public Builder setConnectionTimeout(long timeout) {
-      _connectionTimeout = timeout;
-      return this;
+    /**
+     * Set timeout in mm for session timeout. When a client becomes partitioned from the metadata
+     * service for more than session timeout, new session will be established.
+     * @param timeout
+     * @return
+     */
+    public B setSessionTimeoutInMillis(long timeout) {
+      _sessionTimeoutInMillis = timeout;
+      return self();
     }
 
-    public Builder setStoreType(StoreType storeType) {
+    public B setStoreType(StoreType storeType) {
       _storeType = storeType;
-      return this;
+      return self();
+    }
+
+    @SuppressWarnings("unchecked")
+    final B self() {
+      return (B) this;
     }
 
-    private void validate() {
+    protected void validate() {
```

Comment:
We should add more validations here right? Are you planning to do it in future? I think i can think of is having separate "validate" for each major config classes like "ConnectionConfig", "AuthConfig" etc and make use of them here.

Reply:
Yes we plan to add it later in future.